src/xfastertransformer/tools/llama_convert.py

```diff
@@ -210,3 +210,189 @@ def split_and_convert(self, input_dir, output_dir, dtype, processes):
+        except Exception as e:
+            print("Fail to save the config in config.ini.", str(e))
+
+        hf_model_name_pattern = [
+            "input_layernorm.weight",
+            "self_attn.qkv_proj.qweight",
+            "self_attn.qkv_proj.qzeros",
+            "self_attn.qkv_proj.scales",
+            "self_attn.o_proj.qweight",
+            "self_attn.o_proj.qzeros",
+            "self_attn.o_proj.scales",
+            "post_attention_layernorm.weight",
+            "mlp.gate_proj.qweight",
+            "mlp.gate_proj.qzeros",
+            "mlp.gate_proj.scales",
+            "mlp.up_proj.qweight",
+            "mlp.up_proj.qzeros",
+            "mlp.up_proj.scales",
+            "mlp.down_proj.qweight",
+            "mlp.down_proj.qzeros",
+            "mlp.down_proj.scales",
+        ]
+
+        ft_model_name_pattern = [
+            "input_layernorm.weight",
+            "attention.query_key_value.qweight",
+            "attention.query_key_value.zeros",
+            "attention.query_key_value.scales",
+            "attention.dense.qweight",
+            "attention.dense.zeros",
+            "attention.dense.scales",
+            "post_attention_layernorm.weight",
+            "mlp.gate_proj.qweight",
+            "mlp.gate_proj.zeros",
+            "mlp.gate_proj.scales",
+            "mlp.up_proj.qweight",
+            "mlp.up_proj.zeros",
+            "mlp.up_proj.scales",
+            "mlp.down_proj.qweight",
+            "mlp.down_proj.zeros",
+            "mlp.down_proj.scales",
+        ]
+
+        state_dict = model.state_dict()
+        model_named_parameters = dict()
+        for name, param in state_dict.items():
+            if name.startswith("model."):
+                name = name[6:]
+            wf = torch.tensor(list(range(0, 32, self.wbits)), dtype=torch.int32).unsqueeze(0)
+
+            print(name)
+
+            if "embed" in name:
+                model_named_parameters[name] = param
+            elif "lm_head" in name:
+                model_named_parameters[name] = param
+            elif "scales" in name:
+                model_named_parameters[name] = param.float()
+            elif "qzeros" in name:
+                qzeros = param
+                qzeros = torch.bitwise_right_shift(torch.unsqueeze(qzeros, 2).expand(-1, -1, 32 // self.wbits),
+                        wf.unsqueeze(0)).to(torch.int16 if self.wbits == 8 else torch.int8)
+                qzeros = qzeros + 1
+                qzeros = torch.bitwise_and(qzeros, (2 ** self.wbits) - 1)
+                if self.wbits == 8:
+                    qzeros = qzeros - 2 ** (self.wbits - 1) # uint to int
+                qzeros = torch.flatten(qzeros).float()
+                scales = state_dict["model." + name.replace("qzeros", "scales")].float()
+                zeros = - scales * qzeros
+                model_named_parameters[name] = zeros
+            elif "qweight" in name:
+                # qweight is not transposed
```

Comment:
convert uint8 qweight from QPTQ quantization, and need to convert to int8 for xFT.
convert uint4 qweight from QPTQ quantization, and need to convert to uint4x2 for xFT.